File: src/elixir/test/traceparent_policy_test.exs
defmodule OresMiddleware.TraceparentPolicyTest do
  use ExUnit.Case, async: false

  import Plug.Conn
  import Plug.Test, only: [conn: 2]

  @zero_trace_id String.duplicate("0", 32)
  @valid_trace_id "0123456789abcdef0123456789abcdef"
  @valid_parent_span_id "0123456789abcdef"
  @valid_server_span_id "fedcba9876543210"

  defp stack do
    config = OresMiddleware.default_config("traceparent-policy-test")
    config = put_in(config, [:environment], :test)
    config = put_in(config, [:settings, :tls, :requireHttps], false)
    config = put_in(config, [:settings, :rateLimit, :enabled], false)
    config = put_in(config, [:settings, :idempotency, :enabled], false)
    OresMiddleware.Stack.new!(config)
  end

  defp request(trace_id) do
    conn(:get, "/trace")
    |> put_req_header("accept", "application/json")
    |> put_req_header(
      "traceparent",
      "00-#{trace_id}-#{@valid_parent_span_id}-01"
    )
  end

  test "the inbound parent is not relabelled as a response server span" do
    response =
      OresMiddleware.Plug.wrap(stack(), request(@valid_trace_id), fn conn ->
        resp(conn, 204, "")
      end)

    assert get_resp_header(response, "traceparent") == []
  end

  test "only a valid tracer-owned response traceparent is preserved" do
    valid = "00-#{@valid_trace_id}-#{@valid_server_span_id}-01"

    valid_response =
      OresMiddleware.Plug.wrap(stack(), request(@valid_trace_id), fn conn ->
        conn
        |> put_resp_header("traceparent", String.upcase(valid))
        |> resp(204, "")
      end)

    assert get_resp_header(valid_response, "traceparent") == [valid]

    invalid_response =
      OresMiddleware.Plug.wrap(stack(), request(@valid_trace_id), fn conn ->
        conn
        |> put_resp_header(
          "traceparent",
          "00-#{@valid_trace_id}-0000000000000000-01"
        )
        |> resp(204, "")
      end)

    assert get_resp_header(invalid_response, "traceparent") == []
  end

  test "an all-zero inbound trace ID is replaced" do
    response =
      OresMiddleware.Plug.wrap(stack(), request(@zero_trace_id), fn conn ->
        context = OresMiddleware.current_context()
        assert context.trace_id != @zero_trace_id
        assert context.trace_id =~ ~r/^[0-9a-f]{32}$/
        resp(conn, 204, "")
      end)

    assert response.status == 204
    assert get_resp_header(response, "traceparent") == []
  end
end

File: src/ts/test/traceparent-policy.test.mjs
import test from "node:test";
import assert from "node:assert/strict";

import {
  createMiddleware,
  currentContext,
  defaultConfig
} from "../dist/index.js";

const ZERO_TRACE_ID = "00000000000000000000000000000000";
const VALID_TRACE_ID = "0123456789abcdef0123456789abcdef";
const VALID_PARENT_SPAN_ID = "0123456789abcdef";
const VALID_SERVER_SPAN_ID = "fedcba9876543210";
const VALID_RESPONSE_TRACEPARENT =
  `00-${VALID_TRACE_ID}-${VALID_SERVER_SPAN_ID}-01`;

function testConfig() {
  const config = defaultConfig("traceparent-policy-test");
  config.environment = "test";
  config.settings.tls.mode = "disabled";
  config.settings.tls.requireHttps = false;
  config.settings.rateLimit.enabled = false;
  config.settings.idempotency.enabled = false;
  config.settings.compression.enabled = false;
  return config;
}

function request(traceId = VALID_TRACE_ID) {
  return new Request("http://example.test/trace", {
    headers: {
      accept: "application/json",
      traceparent: `00-${traceId}-${VALID_PARENT_SPAN_ID}-01`
    }
  });
}

test("portable middleware does not echo the inbound parent as a response span", async () => {
  const middleware = createMiddleware(testConfig());
  const response = await middleware(request(), async () =>
    new Response(null, { status: 204 })
  );

  assert.equal(response.headers.get("traceparent"), null);
  assert.ok(response.headers.get("x-request-id"));
});

test("only a valid tracer-owned response traceparent is preserved", async () => {
  const middleware = createMiddleware(testConfig());
  const cases = [
    [VALID_RESPONSE_TRACEPARENT.toUpperCase(), VALID_RESPONSE_TRACEPARENT],
    [`00-${VALID_TRACE_ID}-0000000000000000-01`, null],
    [`00-${ZERO_TRACE_ID}-${VALID_SERVER_SPAN_ID}-01`, null],
    ["00-not-hex-not-a-span-01", null]
  ];

  for (const [candidate, expected] of cases) {
    const response = await middleware(request(), async () =>
      new Response(null, {
        status: 204,
        headers: { traceparent: candidate }
      })
    );
    assert.equal(response.headers.get("traceparent"), expected);
  }
});

test("an all-zero inbound trace ID is replaced and never propagated", async () => {
  const middleware = createMiddleware(testConfig());
  let observedTraceId;

  const response = await middleware(request(ZERO_TRACE_ID), async () => {
    observedTraceId = currentContext()?.traceId;
    return new Response(null, { status: 204 });
  });

  assert.match(observedTraceId, /^[0-9a-f]{32}$/);
  assert.notEqual(observedTraceId, ZERO_TRACE_ID);
  assert.equal(response.headers.get("traceparent"), null);
});
